File: 0053-maximum-subarray/0053-maximum-subarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxSubArray = function(nums) {
    let currentSum = nums[0];
    let maxSum = nums[0];

    for (let i = 1; i < nums.length; i++) {
        // Either take the current element alone or extend the previous subarray
        currentSum = Math.max(nums[i], currentSum + nums[i]);
        maxSum = Math.max(maxSum, currentSum);
    }

    return maxSum;
}
    

File: 0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int a=0;
        int ans=Integer.MIN_VALUE;
        for(int i=0; i<nums.length; i++){
            a+=nums[i];
            ans=Math.max(a,ans);
            if(a<0){
                a=0;
            }
        }return ans;
        
    }
}